apps/pipelines/nodes/nodes.py

```diff
@@ -72,31 +70,119 @@ class CreateReport(LLMResponse):
+        The input should be in the format {"key": "description", "key2": [{"key": "description"}]}
+
+        Nested objects are not supported at the moment
+
+        Input example 1:
+        {"name": "the user's name", "surname": "the user's surname"}
+
+        Input example 2:
+        {"name": "the user's name", "pets": [{"name": "the pet's name": "type": "the type of animal"}]}
+
+        """
+
+        def _create_model_from_data(value_data, model_name: str):
+            pydantic_schema = {}
+            for key, value in value_data.items():
+                if isinstance(value, str):
+                    pydantic_schema[key] = (str | None, Field(description=value))
+                elif isinstance(value, list):
+                    model = _create_model_from_data(value[0], key.capitalize())
+                    pydantic_schema[key] = (list[model], Field(description=f"A list of {key}"))
+            return create_model(model_name, **pydantic_schema)
+
+        Model = _create_model_from_data(data, "CustomModel")
+        schema = Model.model_json_schema()
+        # The schema needs a description in order to comply with function calling APIs
+        schema["description"] = ""
+        return schema
+
+
+class UpdateParticipantMemory(PipelineNode):
+    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
+    the input data will be merged with `key_name` as the key name.
+    """
+
+    __human_name__ = "Update participant memory"
+    key_name: str | None = None
+
+    def _process(self, state: PipelineState) -> PipelineState:
+        """Strategy:
+        - If the input is a string or list, a `key_name` must be present to be valid JSON.
+        - If `key_name` is present, we create a new dictionary with `input` as the value to `key_name`.
+        - Merge this new dictionary with the current participant data, essentially overriding existing data
+        """
+        new_data = state["messages"][-1]
+        session = state["experiment_session"]
+
+        if isinstance(new_data, str) or isinstance(new_data, list):
+            if not self.key_name:
+                raise KeyError("A key is expected for a string or list value.")
+
+        try:
+            participant_data = ParticipantData.objects.for_experiment(session.experiment).get(
+                participant=session.participant
+            )
 
-        return RunnableLambda(fn, name=self.__class__.__name__)
+            if self.key_name:
+                new_data = {self.key_name: new_data}
+
+            participant_data.data = participant_data.data | new_data
+            participant_data.save()
+        except ParticipantData.DoesNotExist:
+            ParticipantData.objects.create(
+                participant=session.participant,
+                content_object=session.experiment,
+                team=session.team,
+                data={self.key_name: new_data} if self.key_name else new_data,
```

Comment:
new_data has already been made into a dict by this stage so this should just be:
```suggestion
                data=new_data,
```

Reply:
https://github.com/dimagi/open-chat-studio/pull/474/commits/bc731e6e2d699a35edb6832d8e761cd97647b5ed